### ISSUE TITLE
Massive Performance Regression with Nested Groups

### ISSUE THREAD

**Issue by jasongilanfarr** [mesosphere/marathon#4813]
When testing scale using nested groups (I had been wondering what was different in my test) against the simulator using the following batch script:

```
#!/bin/bash

for i in {1..500000}; do
http POST :8888/v2/apps <<EOF
{
  "id" : "/$(($i % 17))/$(($i % 13))/$(($i % 11))/$(($i % 7))/simple-app-$i",
  "cmd": "sleep 1000",
  "instances": 1,
  "cpus": 0.01,
  "mem": 1,
  "disk": 0
}
EOF
done
```

I could only get to 154 apps/tasks before marathon couldn't do anything (without nested I was able to get easily over 1000)

The profiler revealed that `Group.transitiveGroups` and `RootGroup.updateVersion` combined together consumed roughly *1.6 GB of Maps in just 1 minute*.

We should probably fully revert the `RootGroup` change in both master and releases/1.4

Related: #4802

**Comment by mpark** [comment]
Documenting my findings here.
We have the following code in `GroupManagerActor.scala`.

```
    dynamicApps.foldLeft(to) { (rootGroup, app) =>
      rootGroup.updateApp(app.id, _ => app, app.version)
    }
```

`updateApp` calls the `Group.group` function which used to traverse the group tree manually.
I had changed `Group.group` to look up the group in `lazy val transitiveGroupsById` instead.

This becomes an issue because `every iteration through the `foldLeft` we construct a new instance
of `RootGroup` from `RootGroup.updateApp`, which subsequently invokes `Group.group`, which
will __recompute__ the `lazy val transitiveGroupsById` since it's a new instance.

That is, we recompute all of the transitive groups for the entire tree on every iteration.

**Comment by mpark** [comment]
Changing the above back to the old definition of `Group`, and removing the uses of `updateVersion`
I was able to get past 1000+ tasks / 1000+ apps.

I need to investigate further as to why `updateVersion` has such a significant impact.

**Comment by mpark** [comment]
I ran the same experiment with my patch reverted on master + D309 + D311, and I only got to 400 ish apps / tasks after 30 mins.

```
There are 377 tasks / 377 apps
There are 378 tasks / 378 apps
There are 378 tasks / 378 apps
There are 378 tasks / 378 apps
There are 378 tasks / 378 apps
There are 378 tasks / 378 apps
There are 379 tasks / 379 apps
There are 379 tasks / 379 apps
There are 379 tasks / 379 apps
There are 379 tasks / 379 apps
There are 379 tasks / 379 apps
There are 380 tasks / 380 apps
There are 380 tasks / 380 apps
There are 380 tasks / 380 apps
There are 380 tasks / 380 apps
There are 380 tasks / 380 apps
There are 381 tasks / 381 apps
There are 381 tasks / 381 apps
There are 381 tasks / 381 apps
There are 381 tasks / 381 apps
There are 381 tasks / 381 apps
There are 381 tasks / 381 apps
There are 381 tasks / 381 apps
There are 381 tasks / 381 apps
There are 381 tasks / 381 apps


```

**Comment by timcharper** [comment]
D320 landed, c8b1742efb8b5f3d65894de153dd4ade4fe774a2 should help reduce # of memory leak due to endless growth of lazy maps referring to the original data structure.

**Comment by mpark** [comment]
https://phabricator.mesosphere.com/D321 was merged and is also very important for this ticket.
